### PR TITLE
selector avalanche protection

### PR DESCRIPTION
**Pull request by LLLWD**
detail：
https://github.com/alibaba/nacos/issues/10178
是这样重新提交一个plz merge吗（还是把close的open，第一次在开源项目上提交merge，抱歉）

### CODE REVIEW

**Comment by KomachiSion** [comment]
Please sign CLA.

**Comment by KomachiSion** [comment]
> [CLA assistant check](https://cla-assistant.io/alibaba/nacos?pullRequest=10184) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/nacos?pullRequest=10184) before we can accept your contribution.
> 
> **Zhengxueliang** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/alibaba/nacos?pullRequest=10184) it.

Read the hint please.

**Comment by LLLWD** [comment]
> > [CLA assistant check](https://cla-assistant.io/alibaba/nacos?pullRequest=10184) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/nacos?pullRequest=10184) before we can accept your contribution.
> > **Zhengxueliang** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> > You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/alibaba/nacos?pullRequest=10184) it.
> 
> Read the hint please.

signed, thank u

File: naming/src/main/java/com/alibaba/nacos/naming/utils/ServiceUtil.java [hunk 1]
```diff
@@ -196,14 +196,14 @@ public static ServiceInfo selectInstancesWithHealthyProtection(ServiceInfo servi
...
-            long newHealthyCount = healthyCount;
-            if (originalTotal != allInstances.size()) {
-                for (com.alibaba.nacos.api.naming.pojo.Instance allInstance : allInstances) {
-                    if (allInstance.isHealthy()) {
-                        newHealthyCount++;
-                    }
+            long newHealthyCount = 0L;
+            // if (originalTotal != allInstances.size()) {
+            for (com.alibaba.nacos.api.naming.pojo.Instance allInstance : allInstances) {
+                if (allInstance.isHealthy()) {
```
Comment (pixystone) [inline naming/src/main/java/com/alibaba/nacos/naming/utils/ServiceUtil.java:202]: Please re-compute this count only when `allInstances` changed considering performance.

I think it just missing the `newHealthyCount = 0L;` in the `if` block started at line 200.

**Review by pixystone — CHANGES_REQUESTED** [review]
Need a small change.

File: naming/src/main/java/com/alibaba/nacos/naming/utils/ServiceUtil.java (hunk 1 above)
Comment (LLLWD) [reply naming/src/main/java/com/alibaba/nacos/naming/utils/ServiceUtil.java]: sry,  have changed

**Comment by pixystone** [comment]
LGTM.

Unit tests are welcome.

**Review by pixystone — APPROVED** [review]
LGTM.

Unit tests are welcome.

**Comment by KomachiSion** [comment]
@LLLWD Can you add some Unit test to test the codes, so that others will not make the codes has problem.

**Comment by LLLWD** [comment]
> @LLLWD Can you add some Unit test to test the codes, so that others will not make the codes has problem.

ok, i will do it this week later

**Comment by LLLWD** [comment]
I had no experience with unit testing for java projects before NACOS, so I had some problems adding unit tests. I may need more time to learn how to write java unit tests and lambda syntax.
I broke my promise, sorry

**Comment by KomachiSion** [comment]
current `ServiceUtil` has unit test class `com.alibaba.nacos.naming.utils.ServiceUtilTest`.

But the test case only one case so that the changes not test all codes.

You can reference others unit test in `com.alibaba.nacos.naming.core.v2` and add the unit test for your change.

Of course, if you can add a full unit test for this class will be better.

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**